File: cpp/density.cpp
#include "density.hpp"
#include "riemann.hpp"
#ifdef ADIFF
    #include "externals/ampi_interface_realreverse.cpp"
#endif

//// confirm that make_tuple doesn't create copies
void RCF::primitive(const scalar rho, const scalar rhoU[3], const scalar rhoE, scalar U[3], scalar& T, scalar& p) {
    scalar U2 = 0.;
    for (integer i = 0; i < 3; i++) {
        U[i] = rhoU[i]/rho;
        U2 += U[i]*U[i];
    }
    scalar e = rhoE/rho - 0.5*U2;
    T = e/this->Cv, 
    p = (this->gamma-1)*rho*e;
}

void RCF::conservative(const scalar U[3], const scalar T, const scalar p, scalar& rho, scalar rhoU[3], scalar& rhoE) {
    scalar e = this->Cv * T;
    rho = p/(e*(this->gamma - 1));
    scalar U2 = 0.;
    for (integer i = 0; i < 3; i++) {
        U2 += U[i]*U[i];
        rhoU[i] = rho*U[i];
    }
    rhoE = rho*(e + 0.5*U2);
}

//inline Ref<arr> internalField(arr& phi) {
    //Ref<arr> phiI = SELECT(phi, 0, mesh.nInternalCells);
    //return phiI;
//}

//inline Ref<arr> boundaryField(arr& phi) { 
    //Ref<arr> phiB = SELECT(phi, mesh.nInternalCells, mesh.nGhostCells);
    //return phiB;
//}

//inline arr mu(const arr& T) {
    //return 0*T;//1.4792e-06*T.pow(1.5)/(T+116);
//}
//inline arr kappa(const arr& mu, const arr& T) {
    //return mu*this->Cp/this->Pr;
//}
void RCF::getFlux(const scalar U[3], const scalar T, const scalar p, const uscalar N[3], scalar& rhoFlux, scalar rhoUFlux[3], scalar& rhoEFlux) {
    scalar rho, rhoU[3], rhoE;
    this->conservative(U, T, p, rho, rhoU, rhoE);
    scalar Un = 0.;
    for (integer i = 0; i < 3; i++) {
        Un += U[i]*N[i];
    }
    rhoFlux = rho*Un;
    for (integer i = 0; i < 3; i++) {
        rhoUFlux[i] = rhoU[i]*Un + p*N[i];
    }
    rhoEFlux = (rhoE + p)*Un;
}

void RCF::equation(const arr& rho, const arr& rhoU, const arr& rhoE, arr& drho, arr& drhoU, arr& drhoE) {
    // make decision between 1 and 3 a template
    // optimal memory layout? combine everything?
    //cout << "c++: equation 1" << endl;
    const Mesh& mesh = *this->mesh;

    arr U(mesh.nCells, 3);
    arr T(mesh.nCells);
    arr p(T.shape);
    for (integer i = 0; i < mesh.nInternalCells; i++) {
        this->primitive(rho(i), &rhoU(i), rhoE(i), &U(i), T(i), p(i));
    }
    //cout << "c++: equation 2" << endl;

    this->boundary(this->boundaries[0], U);
    this->boundary(this->boundaries[1], T);
    this->boundary(this->boundaries[2], p);

    arr gradU(mesh.nCells, 3, 3);
    arr gradT(mesh.nCells, 1, 3);
    arr gradp(gradT.shape);
    gradU.zero();
    gradT.zero();
    gradp.zero();
    auto faceUpdate = [&](integer start, integer end, bool neighbour) {
        for (integer i = start; i < end; i++) {
            scalar UF[3], TF, pF;
            this->interpolate->central(U, UF, i);
            this->interpolate->central(T, &TF, i);
            this->interpolate->central(p, &pF, i);
            this->operate->grad(UF, gradU, i, neighbour);
            this->operate->grad(&TF, gradT, i, neighbour);
            this->operate->grad(&pF, gradp, i, neighbour);
        }
    };
    faceUpdate(0, mesh.nInternalFaces, true);
    for (auto& patch: mesh.boundary) {
        auto& patchInfo = patch.second;
        integer startFace, nFaces;
        tie(startFace, nFaces) = mesh.boundaryFaces.at(patch.first);
        integer cellStartFace = mesh.nInternalCells + startFace - mesh.nInternalFaces;
        if (patchInfo.at("type") == "cyclic") {
            faceUpdate(startFace, startFace + nFaces, false);
        } else {
            for (integer i = 0; i < nFaces; i++) {
                integer index = startFace + i;
                integer c = cellStartFace + i;
                this->operate->grad(&U(c), gradU, index, false);
                this->operate->grad(&T(c), gradT, index, false);
                this->operate->grad(&p(c), gradp, index, false);
            }
        }
    }
    faceUpdate(mesh.nLocalFaces, mesh.nFaces, false);
    //cout << "gradU " << gradU.checkNAN() << endl;
    
    //cout << "c++: equation 3" << endl;
    this->boundary(mesh.defaultBoundary, gradU);
    this->boundary(mesh.defaultBoundary, gradT);
    this->boundary(mesh.defaultBoundary, gradp);
    
    drho.zero();
    drhoU.zero();
    drhoE.zero();
    auto faceFluxUpdate = [&](integer start, integer end, bool neighbour) {
        for (integer i = start; i < end; i++) {
            scalar ULF[3], URF[3];
            scalar TLF, TRF;
            scalar pLF, pRF;
            
            this->interpolate->secondOrder(U, gradU, ULF, i, 0);
            this->interpolate->secondOrder(U, gradU, URF, i, 1);
            this->interpolate->secondOrder(T, gradT, &TLF, i, 0);
            this->interpolate->secondOrder(T, gradT, &TRF, i, 1);
            this->interpolate->secondOrder(p, gradp, &pLF, i, 0);
            this->interpolate->secondOrder(p, gradp, &pRF, i, 1);

            scalar rhoLF, rhoRF;
            scalar rhoULF[3], rhoURF[3];
            scalar rhoELF, rhoERF;
            this->conservative(ULF, TLF, pLF, rhoLF, rhoULF, rhoELF);
            this->conservative(URF, TRF, pRF, rhoRF, rhoURF, rhoERF);
            scalar rhoFlux;
            scalar rhoUFlux[3];
            scalar rhoEFlux;
            riemannSolver(this->gamma, \
                pLF, pRF, TLF, TRF, ULF, URF, \
                rhoLF, rhoRF, rhoULF, rhoURF, rhoELF, rhoERF, &mesh.normals(i), 
                rhoFlux, rhoUFlux, rhoEFlux);
            this->operate->div(&rhoFlux, drho, i, neighbour);
            this->operate->div(rhoUFlux, drhoU, i, neighbour);
            this->operate->div(&rhoEFlux, drhoE, i, neighbour);
        }
    };
    faceFluxUpdate(0, mesh.nInternalFaces, true);
    //cout << "c++: equation 4" << endl;
    // characteristic boundary
    for (auto& patch: mesh.boundary) {
        auto& patchInfo = patch.second;
        integer startFace, nFaces;
        tie(startFace, nFaces) = mesh.boundaryFaces.at(patch.first);
        integer cellStartFace = mesh.nInternalCells + startFace - mesh.nInternalFaces;
        if (patchInfo.at("type") == "cyclic") {
            faceFluxUpdate(startFace, startFace + nFaces, false);
        } else {
            for (integer i = 0; i < nFaces; i++) {
                integer index = startFace + i;
                integer c = cellStartFace + i;
                scalar rhoFlux;
                scalar rhoUFlux[3];
                scalar rhoEFlux;
                this->getFlux(&U(c), T(c), p(c), &mesh.normals(index), rhoFlux, rhoUFlux, rhoEFlux);
                this->operate->div(&rhoFlux, drho, index, false);
                this->operate->div(rhoUFlux, drhoU, index, false);
                this->operate->div(&rhoEFlux, drhoE, index, false);
            }
        }
    }
    faceFluxUpdate(mesh.nLocalFaces, mesh.nFaces, false);
    //cout << "c++: equation 5" << endl;
}

void RCF::boundary(const Boundary& boundary, arr& phi) {
    const Mesh& mesh = *this->mesh;
    //MPI_Barrier(MPI_COMM_WORLD);

    arr phiBuf(mesh.nCells-mesh.nLocalCells, phi.shape[1], phi.shape[2]);
    AMPI_Request* req;
    integer reqIndex = 0;
    if (mesh.nRemotePatches > 0) {
        req = new AMPI_Request[2*mesh.nRemotePatches];
    }

    for (auto& patch: boundary) {
        string patchType = patch.second.at("type");
        string patchID = patch.first;
        const map<string, string>& patchInfo = mesh.boundary.at(patchID);

        integer startFace, nFaces;
        tie(startFace, nFaces) = mesh.boundaryFaces.at(patch.first);
        integer cellStartFace = mesh.nInternalCells + startFace - mesh.nInternalFaces;

        if (patchType == "cyclic") {
            string neighbourPatchID = patchInfo.at("neighbourPatch");
            integer neighbourStartFace = std::get<0>(mesh.boundaryFaces.at(neighbourPatchID));
            for (integer i = 0; i < nFaces; i++) {
                integer p = mesh.owner(neighbourStartFace + i);
                integer c = cellStartFace + i;
                for (integer j = 0; j < phi.shape[1]; j++) {
                    for (integer k = 0; k < phi.shape[2]; k++) {
                        phi(c, j, k) = phi(p, j, k);
                    }
                }
            }
        } else if (patchType == "zeroGradient" || patchType == "empty" || patchType == "inletOutlet") {
            for (integer i = 0; i < nFaces; i++) {
                integer p = mesh.owner(startFace + i);
                integer c = cellStartFace + i;
                for (integer j = 0; j < phi.shape[1]; j++) {
                    for (integer k = 0; k < phi.shape[2]; k++) {
                        phi(c, j, k) = phi(p, j, k);
                    }
                }
            }
        } else if (patchType == "symmetryPlane" || patchType == "slip") {
            if ((phi.shape[1] == 3) && (phi.shape[2] == 1)) {
                for (integer i = 0; i < nFaces; i++) {
                    integer f = startFace + i;
                    integer c = cellStartFace + i;
                    integer p = mesh.owner(f);
                    scalar phin = 0.;
                    for (integer j = 0; j < 3; j++) {
                        phin += mesh.normals(f, j)*phi(p, j);
                    }
                    for (integer j = 0; j < 3; j++) {
                        phi(c, j) = phi(p, j) - mesh.normals(f, j)*phin;
                    }
                }
            } else {
                for (integer i = 0; i < nFaces; i++) {
                    integer p = mesh.owner(startFace + i);
                    integer c = cellStartFace + i;
                    for (integer j = 0; j < phi.shape[1]; j++) {
                        for (integer k = 0; k < phi.shape[2]; k++) {
                            phi(c, j, k) = phi(p, j, k);
                        }
                    }
                }
            }
        } else if (patchType == "fixedValue") {
            // fix this
            // more boundary conditions
            if (phi.shape[1] == 3) {
                for (integer i = 0; i < nFaces; i++) {
                    phi(cellStartFace + i, 0) = 3.;
                    phi(cellStartFace + i, 1) = 0.;
                    phi(cellStartFace + i, 2) = 0.;
                } 
            } else {
                for (integer i = 0; i < nFaces; i++) {
                    phi(cellStartFace + i) = 1.;
                }
            }
        } else if (patchType == "processor" || patchType == "processorCyclic") {
            //cout << "hello " << patchID << endl;
            integer bufStartFace = cellStartFace - mesh.nLocalCells;
            integer size = nFaces*phi.shape[1]*phi.shape[2];
            integer dest = stoi(patchInfo.at("neighbProcNo"));
            integer tag = 0;
            for (integer i = 0; i < nFaces; i++) {
                integer p = mesh.owner(startFace + i);
                integer b = bufStartFace + i;
                for (integer j = 0; j < phi.shape[1]; j++) {
                    for (integer k = 0; k < phi.shape[2]; k++) {
                        phiBuf(b, j, k) = phi(p, j, k);
                    }
                }
            }
            AMPI_Isend(&phiBuf(bufStartFace), size, MPI_DOUBLE, dest, tag, MPI_COMM_WORLD, &req[reqIndex]);
            AMPI_Irecv(&phi(cellStartFace), size, MPI_DOUBLE, dest, tag, MPI_COMM_WORLD, &req[reqIndex+1]);
            reqIndex += 2;
        }
        else {
            cout << "patch not found " << patchType << " for " << patchID << endl;
        }
    }
    if (mesh.nRemotePatches > 0) {
        AMPI_Waitall(2*mesh.nRemotePatches, req, MPI_STATUSES_IGNORE);
        delete[] req;
    }
}

File: adFVM/density.py
from . import config, riemann, interp
from .field import Field, IOField, CellField
from .op import  div, absDiv, snGrad, grad, gradCell, snGradCorr
from .solver import Solver
from .interp import central
from .interp import secondOrder as faceInterpolator
from . import BCs
from . import postpro
from . import timestep
from .mesh import Mesh

from adpy.tensor import Kernel, ExternalFunctionOp
from adpy.variable import Variable, Function, Zeros

import numpy as np
#import adFVMcpp

logger = config.Logger(__name__)


class RCF(Solver):
    defaultConfig = Solver.defaultConfig.copy()
    defaultConfig.update({
                             #specific
                             'Cp': 1004.5, 
                             'gamma': 1.4, 
                             'mu': lambda T:  1.4792e-06*T**1.5/(T+116.), 
                             'Pr': 0.7, 
                             'CFL': 1.2,
                             'stepFactor': 1.2,
                             'timeIntegrator': 'SSPRK', 'nStages': 3,
                             # eulerHLLC DOES NOT WORK
                             'riemannSolver': 'eulerRoe',
                             'objectivePLInfo': None,
                             #'boundaryRiemannSolver': 'eulerLaxFriedrichs',
                             'boundaryRiemannSolver': 'eulerRoe',
                             'readConservative': False,
                             'faceReconstructor': 'secondOrder',
                        })

    def __init__(self, case, **userConfig):
        super(RCF, self).__init__(case, **userConfig)

        self.Cv = self.Cp/self.gamma
        self.R = self.Cp - self.Cv
        self.kappa = lambda mu, T: mu*(self.Cp/self.Pr)
        #self.kappa = lambda mu, T: 0.
        self.riemannSolver = getattr(riemann, self.riemannSolver)
        self.boundaryRiemannSolver = getattr(riemann, self.boundaryRiemannSolver)
        self.faceReconstructor = getattr(interp, self.faceReconstructor)

        self.names = ['rho', 'rhoU', 'rhoE']
        self.dimensions = [(1,), (3,), (1,)]

        self.initSource()

        self.Uref = 33.
        self.Tref = 300.
        self.pref = 1e5
        self.tref = 1.
        self.Jref = 1.
        return

    def compileInit(self):
        mesh = self.mesh.symMesh
        self._primitive = Kernel(self.primitive)
        self._conservative = Kernel(self.conservative)
        meshArgs = mesh.getTensor() + mesh.getScalar()
        BCArgs = self.getBoundaryTensor(0)
        rho, rhoU, rhoE = Variable((mesh.nInternalCells, 1)), Variable((mesh.nInternalCells, 3)), Variable((mesh.nInternalCells, 1)),
        U, T, p = Zeros((mesh.nCells, 3)), Zeros((mesh.nCells, 1)), Zeros((mesh.nCells, 1))
        outputs = tuple([x.getReference() for x in [U, T, p]])
        U, T, p = self._primitive(mesh.nInternalCells, outputs)(rho, rhoU, rhoE)
        U, T, p = self.boundaryInit(U, T, p)
        U, T, p = self.boundary(U, T, p)
        U, T, p = self.boundaryEnd(U, T, p)
        rhoN, rhoUN, rhoEN = self._conservative()(U, T, p)
        self.mapBoundary = Function('init', [rho, rhoU, rhoE] + meshArgs + BCArgs, [rhoN, rhoUN, rhoEN])

    def compileSolver(self):
        self._grad = Kernel(self.gradients)
        self._gradCell = Kernel(self.gradientsCell)
        self._coupledGrad = Kernel(self.gradients)
        self._boundaryGrad = Kernel(self.gradients)
        self._flux = Kernel(self.flux)
        self._coupledFlux = Kernel(self.flux)
        self._characteristicFlux = Kernel(self.flux)
        self._boundaryFlux = Kernel(self.boundaryFlux)
        mesh = self.mesh.symMesh
        meshArgs = mesh.getTensor() + mesh.getScalar()
        sourceArgs = [x[0] for x in self.sourceTerms]
        BCArgs = self.getBoundaryTensor(0)
        extraArgs = [x[0] for x in self.extraArgs]
        # init function
        
        #self.t0 = Variable((1,1))
        self.dt = Variable((1,1))
        #self.t = self.t0
        rho, rhoU, rhoE = Variable((mesh.nInternalCells, 1)), Variable((mesh.nInternalCells, 3)), Variable((mesh.nInternalCells, 1)),
        rhoN, rhoUN, rhoEN = timestep.timeStepper(self.equation, [rho, rhoU, rhoE], self)
        args = [rho, rhoU, rhoE, self.dt] + meshArgs + sourceArgs + BCArgs + extraArgs
        outputs = [rhoN, rhoUN, rhoEN, self.dtc, self.obj]
        io_map = {0: 0, 1:1, 2:2}
        #io_map = {}
        self.map = Function('primal', args, outputs, io_map=io_map)
    
        
    def getBoundaryTensor(self, index=0):
        return super(RCF, self).getBoundaryTensor(index) + \
               sum([phi.getTensor(index) for phi in self.gradFields], [])

    # only reads fields
    @config.timeFunction('Time for reading fields')
    def readFields(self, t, suffix=''):
        with IOField.handle(t):
            U = IOField.read('U' + suffix, skipField=False)
            T = IOField.read('T' + suffix, skipField=False)
            p = IOField.read('p' + suffix, skipField=False)
            if self.readConservative:
                rho = IOField.read('rho' + suffix, skipField=False)
                rhoU = IOField.read('rhoU' + suffix, skipField=False)
                rhoE = IOField.read('rhoE' + suffix, skipField=False)
                U.field, T.field, p.field = [phi.field for phi in self.primitive(rho, rhoU, rhoE)]
            if self.dynamicMesh:
                self.mesh.read(IOField._handle)
        fields = [U, T, p]
        # IO Fields, with phi attribute as a CellField
        if self.firstRun:
            self.U, self.T, self.p = fields
            self.fields = fields
            for phi in self.fields:
                phi.completeField()
            self.gradFields = [CellField(phi.name, None, phi.dimensions + (3,)) for phi in self.fields]
        else:
            self.updateFields(fields)
        self.firstRun = False
        return list(self.conservative(*self.fields))

    # reads and updates ghost cells
    def initFields(self, fields):
        mesh = self.mesh
        newFields = self.mapBoundary(*[phi.field for phi in fields] + mesh.getTensor() + mesh.getScalar() + self.getBoundaryTensor(1))
        return self.getFields(newFields, IOField, refFields=fields)
    
    @config.timeFunction('Time for writing fields')
    def writeFields(self, fields, t):
        n = len(self.names)
        fields, rest = fields[:n], fields[n:]
        fields = self.initFields(fields)
        U, T, p = self.primitive(*fields)
        for phi, phiN in zip(self.fields, [U, T, p]):
            phi.field = phiN.field

        with IOField.handle(t):
            for phi in fields + self.fields + rest:
                phi.write(skipProcessor=True)
            if self.dynamicMesh:
                self.mesh.write(IOField._handle)
        return

    # operations, dual
    def primitive(self, rho, rhoU, rhoE):
        logger.info('converting fields to primitive')
        U = rhoU/rho
        E = rhoE/rho
        e = E - 0.5*U.magSqr()
        p = (self.gamma-1)*rho*e
        T = e*(1./self.Cv)
        if isinstance(U, Field):
            U.name, T.name, p.name = 'U', 'T', 'p'
        return U, T, p

    def conservative(self, U, T, p):
        logger.info('converting fields to conservative')
        e = self.Cv*T
        rho = p/(e*(self.gamma-1))
        rhoE = rho*(e + 0.5*U.magSqr())
        rhoU = U*rho
        if isinstance(U, Field):
            rho.name, rhoU.name, rhoE.name = self.names
        return rho, rhoU, rhoE

    # used in symbolic
    def getFlux(self, U, T, p, Normals):
        rho, rhoU, rhoE = self.conservative(U, T, p)
        Un = U.dot(Normals)
        rhoFlux = rho*Un
        rhoUFlux = rhoU*Un + p*Normals
        rhoEFlux = (rhoE + p)*Un
        return rhoFlux, rhoUFlux, rhoEFlux

    def viscousFlux(self, TL, TR, UL, UR, TF, UF, gradTF, gradUF, mesh):
        mu = self.mu(TF)
        kappa = self.kappa(mu, TF)
        D = mesh.deltasUnit
        N = mesh.normals

        # charles
        gradTF = gradTF[0] + snGrad(TL, TR, mesh)*D - gradTF[0].dot(D)*D
        gradUF =  gradUF + snGrad(UL, UR, mesh).outer(D) - gradUF.tensordot(D).outer(D)
        # stable, too dissipative
        #gradTF = gradTF[0]
        #gradUF = gradUF 
        qF = kappa*gradTF.dot(N)
        tmp2 = (gradUF + gradUF.transpose()).tensordot(N)

        # no snGrad correction
        #qF = kappa*snGrad(TL, TR, mesh)
        #tmp2 = snGrad(UL, UR, mesh) + gradUF.transpose().tensordot(N)
        # snGrad correction 
        #qF = kappa*snGradCorr(TL, TR, gradTF, mesh)
        #tmp2 = snGradCorr(UL, UR, gradUF, mesh) + gradUF.transpose().tensordot(N)
        # old theano setup
        #qF = kappa*snGrad(TL, TR, mesh)
        #tmp2 = (gradUF + gradUF.transpose()).tensordot(N)

        tmp3 = gradUF.trace()
        sigmaF = mu*(tmp2-2./3*tmp3*N)

        rhoUFlux = -sigmaF
        rhoEFlux = -(qF + sigmaF.dot(UF))
        return rhoUFlux, rhoEFlux

    #symbolic funcs
    def gradients(self, U, T, p, *mesh, **options):
        mesh = Mesh.container(mesh)
        neighbour = options.pop('neighbour', True)
        boundary = options.pop('boundary', False)
        if boundary:
            UF = U.extract(mesh.neighbour)
            TF = T.extract(mesh.neighbour)
            pF = p.extract(mesh.neighbour)
        else:
            UF = central(U, mesh)
            TF = central(T, mesh)
            pF = central(p, mesh)

        gradU = grad(UF, mesh, neighbour)
        gradT = grad(TF, mesh, neighbour)
        gradp = grad(pF, mesh, neighbour)

        return gradU, gradT, gradp


    def gradientsCell(self, U, T, p, *mesh):
        mesh = Mesh.container(mesh)
        gradU = gradCell(U, mesh)
        gradT = gradCell(T, mesh)
        gradp = gradCell(p, mesh)
        
        return gradU, gradT, gradp
        
    def flux(self, U, T, p, gradU, gradT, gradp, *mesh, **options):
        mesh = Mesh.container(mesh)
        neighbour = options.pop('neighbour', True)
        characteristic = options.pop('characteristic', False)
        P, N = mesh.owner, mesh.neighbour

        ULF = faceInterpolator(U, gradU, mesh, 0)
        TLF = faceInterpolator(T, gradT, mesh, 0)
        pLF = faceInterpolator(p, gradp, mesh, 0)
        rhoLF, rhoULF, rhoELF = self.conservative(ULF, TLF, pLF)
        TL, TR = T.extract(P), T.extract(N)
        UL, UR = U.extract(P), U.extract(N)

        if characteristic:
            URF, TRF, pRF = UR, TR, p.extract(N)
            rhoRF, rhoURF, rhoERF = self.conservative(URF, TRF, pRF)
            rhoFlux, rhoUFlux, rhoEFlux = self.boundaryRiemannSolver(self.gamma, pLF, pRF, TLF, TRF, ULF, URF, \
            rhoLF, rhoRF, rhoULF, rhoURF, rhoELF, rhoERF, mesh.normals)
            TF = TR
            UF = UR
            gradTF = gradT.extract(N)
            gradUF = gradU.extract(N)
        else:
            URF = faceInterpolator(U, gradU, mesh, 1)
            TRF = faceInterpolator(T, gradT,  mesh, 1)
            pRF = faceInterpolator(p, gradp, mesh, 1)
            rhoRF, rhoURF, rhoERF = self.conservative(URF, TRF, pRF)
            rhoFlux, rhoUFlux, rhoEFlux = self.riemannSolver(self.gamma, pLF, pRF, TLF, TRF, ULF, URF, \
            rhoLF, rhoRF, rhoULF, rhoURF, rhoELF, rhoERF, mesh.normals)
            # more accurate
            UF = 0.5*(ULF + URF)
            TF = 0.5*(TLF + TRF)
            #gradUF = central(gradU, mesh)
            #gradTF = central(gradT, mesh)
            # charles
            #TF = 0.5*(TL + TR)
            #UF = 0.5*(UL + UR)
            gradTF = 0.5*(gradT.extract(P) + gradT.extract(N))
            gradUF = 0.5*(gradU.extract(P) + gradU.extract(N))

        ret = self.viscousFlux(TL, TR, UL, UR, TF, UF, gradTF, gradUF, mesh)
        rhoUFlux += ret[0]
        rhoEFlux += ret[1]

        drho = div(rhoFlux, mesh, neighbour)
        drhoU = div(rhoUFlux, mesh, neighbour)
        drhoE = div(rhoEFlux, mesh, neighbour)

        aF = (self.Cp*TF*(self.gamma-1)).sqrt()
        maxaF = abs(UF.dot(mesh.normals)) + aF
        dtc = absDiv(maxaF, mesh, neighbour)

        return drho, drhoU, drhoE, dtc

    def boundaryFlux(self, U, T, p, gradU, gradT, gradp, *mesh):
        mesh = Mesh.container(mesh)
        P, N = mesh.owner, mesh.neighbour

        # boundary extraction could be done using cellstartface
        UR, TR, pR = U.extract(N), T.extract(N), p.extract(N) 
        #UR, TR, pR = U, T, p
        gradUR, gradTR = gradU.extract(N), gradT.extract(N)
        TL = T.extract(P)
        UL = U.extract(P)

        rhoFlux, rhoUFlux, rhoEFlux = self.getFlux(UR, TR, pR, mesh.normals)
        ret = self.viscousFlux(TL, TR, UL, UR, TR, UR, gradTR, gradUR, mesh)
        rhoUFlux += ret[0]
        rhoEFlux += ret[1]

        drho = div(rhoFlux, mesh, False)
        drhoU = div(rhoUFlux, mesh, False)
        drhoE = div(rhoEFlux, mesh, False)

        aF = (self.Cp*TR*(self.gamma-1)).sqrt()
        maxaF = abs(UR.dot(mesh.normals)) + aF
        dtc = absDiv(maxaF, mesh, False)

        return drho, drhoU, drhoE, dtc

    def characteristicBoundary(self, U, T, p):
        for patchID, patch in self.p.phi.BC.items():
            if isinstance(patch, BCs.CharacteristicBoundaryCondition):
                U, T, p = patch.update(U, T, p)
        return U, T, p

    def equation(self, rho, rhoU, rhoE):
        logger.info('computing RHS/LHS')
        mesh = self.mesh.symMesh

        def _meshArgs(start=0):
            return [x[start] for x in mesh.getTensor()]

        U, T, p = Zeros((mesh.nCells, 3)), Zeros((mesh.nCells, 1)), Zeros((mesh.nCells, 1))
        outputs = self._primitive(mesh.nInternalCells, (U, T, p))(rho, rhoU, rhoE)
        # boundary update
        outputs = self.boundaryInit(*outputs)
        #is this right?
        outputs = self.boundary(*outputs)
        outputs = self.boundaryEnd(*outputs)
        U, T, p = outputs
        if self.objective is not None:
            obj = self.objective([U, T, p], self)
        else:
            obj = Zeros((1,1))

        meshArgs = _meshArgs()
        gradU, gradT, gradp = Zeros((mesh.nCells, 3, 3)), Zeros((mesh.nCells, 1, 3)), Zeros((mesh.nCells, 1, 3))

        #outputs = self._grad(mesh.nInternalFaces, (gradU, gradT, gradp))(U, T, p, *meshArgs)
        #for patchID in self.mesh.sortedPatches:
        #    startFace, nFaces = mesh.boundary[patchID]['startFace'], mesh.boundary[patchID]['nFaces']
        #    patchType = self.mesh.boundary[patchID]['type']
        #    meshArgs = _meshArgs(startFace)
        #    if patchType in config.coupledPatches:
        #        outputs = self._coupledGrad(nFaces, outputs)(U, T, p, neighbour=False, boundary=False, *meshArgs)
        #        outputs[0].args[0].info += [[x.shape for x in self.mesh.getTensor()], patchID, self.mesh.boundary[patchID]['startFace'], self.mesh.boundary[patchID]['nFaces']]
        #    else:
        #        outputs = self._boundaryGrad(nFaces, outputs)(U, T, p, neighbour=False, boundary=True, *meshArgs)
        #meshArgs = _meshArgs(mesh.nLocalFaces)
        #outputs = self._coupledGrad(mesh.nRemoteCells, outputs)(U, T, p, neighbour=False, boundary=False, *meshArgs)
        outputs = self._gradCell(mesh.nInternalCells, (gradU, gradT, gradp))(U, T, p, *meshArgs)

        # grad boundary update
        outputs = list(self.boundaryInit(*outputs))
        for index, phi in enumerate(outputs):
            phi = self.gradFields[index].updateGhostCells(phi)
            #if not config.gpu:
            #    phi = ExternalFunctionOp('mpi', (phi,), (phi,)).outputs[0]
            phi = ExternalFunctionOp('mpi', (phi,), (phi,)).outputs[0]
            outputs[index] = phi
        outputs = self.boundaryEnd(*outputs)
        gradU, gradT, gradp = outputs
        
        meshArgs = _meshArgs()
        drho, drhoU, drhoE = Zeros((mesh.nInternalCells, 1)), Zeros((mesh.nInternalCells, 3)), Zeros((mesh.nInternalCells, 1))
        dtc = Zeros((mesh.nInternalCells, 1))
        outputs = self._flux(mesh.nInternalFaces, (drho, drhoU, drhoE, dtc))(U, T, p, gradU, gradT, gradp, *meshArgs)
        for patchID in self.mesh.sortedPatches:
            startFace, nFaces = mesh.boundary[patchID]['startFace'], mesh.boundary[patchID]['nFaces']
            patchType = self.mesh.boundary[patchID]['type']
            meshArgs = _meshArgs(startFace)
            if patchType in config.coupledPatches:
                outputs = self._coupledFlux(nFaces, outputs)(U, T, p, gradU, gradT, gradp, characteristic=False, neighbour=False, *meshArgs)
            elif patchType == 'characteristic':
                outputs = self._characteristicFlux(nFaces, outputs)(U, T, p, gradU, gradT, gradp, characteristic=True, neighbour=False, *meshArgs)
            else:
                outputs = self._boundaryFlux(nFaces, outputs)(U, T, p, gradU, gradT, gradp, *meshArgs)
        meshArgs = _meshArgs(mesh.nLocalFaces)
        outputs = self._coupledFlux(mesh.nRemoteCells, outputs)(U, T, p, gradU, gradT, gradp, characteristic=False, neighbour=False, *meshArgs)
        drho, drhoU, drhoE, dtc = outputs

        def _minDtc(dtc):
            #dtc = (dtc**-1)*(2*self.CFL)
            #return dtc.reduce_min()
            return dtc.reduce_max()
        minDtc = Zeros((1, 1))
        minDtc = Kernel(_minDtc)(mesh.nInternalCells, (minDtc,))(dtc)

        if self.stage == 1:
            self.dtc, self.obj = minDtc, obj
        return drho, drhoU, drhoE

    def boundary(self, U, T, p):
        outputs = super(RCF, self).boundary(U, T, p, boundary=[phi.phi for phi in self.fields])
        return self.characteristicBoundary(*outputs)
